File: apps/app/modules/pollycar-map/android/src/main/java/expo/modules/pollycarmap/PollyCarMapModule.kt
package expo.modules.pollycarmap

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class PollyCarMapModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("PollyCarMap")

    Constants {
      mapOf(
        "provider" to if (PollyCarMapView.isAmapAvailable()) "amap" else "synthetic",
        "gates" to mapOf(
          "realMapEnabled" to PollyCarMapView.isAmapAvailable(),
          "externalMapProviderEnabled" to PollyCarMapView.isAmapAvailable(),
          "realDeviceLocationEnabled" to false,
          "backgroundLocationEnabled" to false,
          "realVehicleLocationStreamEnabled" to false,
          "amapSdkEnabled" to PollyCarMapView.isAmapAvailable(),
          "amapWebServiceEnabled" to false
        )
      )
    }

    AsyncFunction("initializePrivacy") { state: Map<String, Any?> ->
      PollyCarMapView.updatePrivacy(
        requireNotNull(appContext.reactContext),
        state["noticeContainsAmapPolicy"] == true,
        state["noticeShown"] == true,
        state["consentGranted"] == true
      )
    }

    AsyncFunction("readDeviceLocation") {
      throw IllegalStateException("REAL_DEVICE_LOCATION_UNAVAILABLE")
    }

    AsyncFunction("setBackgroundLocationEnabled") { enabled: Boolean ->
      if (enabled) throw IllegalStateException("BACKGROUND_LOCATION_DISABLED")
    }

    View(PollyCarMapView::class) {
      Events("onCameraIdle", "onMapPress")
      Prop("centerLatitude") { view, latitude: Double -> view.setCenter(latitude = latitude) }
      Prop("centerLongitude") { view, longitude: Double -> view.setCenter(longitude = longitude) }
      Prop("zoom") { view, zoom: Double -> view.setZoom(zoom) }
      Prop("interactive") { view, enabled: Boolean -> view.setInteractive(enabled) }
      Prop("showsUserLocation") { view, enabled: Boolean -> view.setShowsUserLocation(enabled) }
      Prop("markersJson") { view, markersJson: String -> view.setMarkers(markersJson) }
    }
  }
}

File: apps/app/modules/pollycar-map/ios/PollyCarMapModule.swift
import ExpoModulesCore

#if canImport(AMapFoundationKit)
import AMapFoundationKit
#endif

#if canImport(MAMapKit)
import MAMapKit
#endif

public final class PollyCarMapModule: Module {
  public func definition() -> ModuleDefinition {
    Name("PollyCarMap")

    Constants {
      let available = PollyCarMapView.isAmapAvailable
      return [
        "provider": available ? "amap" : "synthetic",
        "gates": [
          "realMapEnabled": available,
          "externalMapProviderEnabled": available,
          "realDeviceLocationEnabled": false,
          "backgroundLocationEnabled": false,
          "realVehicleLocationStreamEnabled": false,
          "amapSdkEnabled": available,
          "amapWebServiceEnabled": false
        ]
      ]
    }

    AsyncFunction("initializePrivacy") { (state: [String: Any]) in
      guard
        state["noticeContainsAmapPolicy"] as? Bool == true,
        state["noticeShown"] as? Bool == true,
        state["consentGranted"] as? Bool == true
      else {
        throw NSError(
          domain: "PollyCarMap",
          code: 3,
          userInfo: [NSLocalizedDescriptionKey: "AMAP_PRIVACY_CONSENT_REQUIRED"]
        )
      }
      #if canImport(AMapFoundationKit) && canImport(MAMapKit)
      guard
        let apiKey = Bundle.main.object(forInfoDictionaryKey: "PollyCarAmapApiKey") as? String,
        !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
      else {
        throw NSError(
          domain: "PollyCarMap",
          code: 4,
          userInfo: [NSLocalizedDescriptionKey: "AMAP_IOS_API_KEY_REQUIRED"]
        )
      }
      AMapServices.shared().apiKey = apiKey
      MAMapView.updatePrivacyShow(.didShow, privacyInfo: .didContain)
      MAMapView.updatePrivacyAgree(.didAgree)
      #else
      throw NSError(
        domain: "PollyCarMap",
        code: 5,
        userInfo: [NSLocalizedDescriptionKey: "AMAP_IOS_SDK_UNAVAILABLE"]
      )
      #endif
    }
    AsyncFunction("readDeviceLocation") {
      throw NSError(domain: "PollyCarMap", code: 1, userInfo: [NSLocalizedDescriptionKey: "REAL_DEVICE_LOCATION_UNAVAILABLE"])
    }
    AsyncFunction("setBackgroundLocationEnabled") { (enabled: Bool) in
      if enabled {
        throw NSError(domain: "PollyCarMap", code: 2, userInfo: [NSLocalizedDescriptionKey: "BACKGROUND_LOCATION_DISABLED"])
      }
    }

    View(PollyCarMapView.self) {
      Events("onCameraIdle", "onMapPress")
      Prop("centerLatitude") { view, latitude: Double in view.setCenter(latitude: latitude) }
      Prop("centerLongitude") { view, longitude: Double in view.setCenter(longitude: longitude) }
      Prop("zoom") { view, zoom: Double in view.setZoom(zoom) }
      Prop("interactive") { view, enabled: Bool in view.setInteractive(enabled) }
      Prop("showsUserLocation") { view, enabled: Bool in view.setShowsUserLocation(enabled) }
      Prop("markersJson") { view, markersJson: String in view.setMarkers(markersJson) }
    }
  }
}
